restart_sampling.py

```diff
@@ -33,7 +39,25 @@ def resolve_t_value(val, ms):
     raise ValueError("bad t_min or t_max value")
 
 
-def prepare_restart_segments(restart_info, ms):
+def prepare_restart_segments(restart_info, ms, sigmas):
+    restart_info = restart_info.strip().lower()
+    if restart_info == "":
+        # No restarts.
+        return []
+    if restart_info == "default":
+        restart_info = DEFAULT_SEGMENTS
+    elif restart_info == "a1111":
+        # Emulate A1111 WebUI's restart sampler behavior.
+        steps = len(sigmas) - 1
+        if steps < 20:
+            # Less than 20 steps - no restarts.
+            return []
+        if steps < 36:
+            # Less than 36 steps - one restart with 9 steps.
+            restart_info = "[10,1,0.1,0.2]"
+        else:
+            # Otherwise two restarts with steps // 4 steps.
+            restart_info = f"[{(steps // 4) + 1}, 2, 0.1, 0.2]"
```

Comment:
ahh, you're right. i did mess that up! thanks for catching it. 

it looks like they're also fixing up the key part (t_min here) with `int(torch.argmin(abs(sigmas - key), dim=0))`, so would i also need to use `int(torch.argmin(abs(sigmas - 0.1), dim=0))` instead there? https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/bef51aed032c0aaa5cfd80445bc4cf0d85b408b5/modules/sd_samplers_extra.py#L50

i don't fully understand what that part is doing.

Reply:
So my understanding is that `max_idx = int(torch.argmin(abs(sigmas - restart_max), dim=0))` is finding the index of the closest sigma to restart_max, which would be 2. 
Then it is used in `sigma_restart = get_sigmas_karras(restart_steps, sigmas[min_idx].item(), sigmas[max_idx].item(), device=sigmas.device)[:-1]`, so ` sigmas[max_idx].item()` is the sigma in the original schedule closest to 2.
So how we have it now, only the `sigma_min`s are rounded to the nearest `sigma` in the schedule as per the paper, and it seems to me that the version in A1111 is rounding their `sigma_max`s to the nearest `sigma`s too.
I'm guessing we can do the same by replacing 2 with the nearest sigma to 2.
Correct me if I'm wrong.

Reply:
seems like my initial naive understanding from reading that code was pretty far off. :( sorry about that.

wouldn't we want to choose the closest sigmas for both `0.1` and `0.2` since it's using that calculation in both cases?

Reply:
We already select the closest sigma to 0.1 since it `s_min` rounded to the nearest sigmas in `round_restart_segments`, but we don't do it with `s_max`, which is what it seems like A1111's algorithm does.

Reply:
makes sense. how about this? hopefully i got it right this time!